sklearn/cross_validation.py

```diff
@@ -1185,8 +1185,10 @@ def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params):
     # Adjust length of sample weights
     n_samples = _num_samples(X)
     fit_params = fit_params if fit_params is not None else {}
-    fit_params = dict([(k, np.asarray(v)[train]
-                       if hasattr(v, '__len__') and len(v) == n_samples else v)
+    fit_params = dict([(k, (safe_indexing(v.tocsr(), train) if sp.issparse(v)
+                            else np.asarray(v)[train])
+                        if _is_arraylike(v) and _num_samples(v) == n_samples
+                        else v)
```

Comment:
Actually the end of this expression is weird. Shouldn't it be `else (k, b)` instead of `else v`? I this case properly tested?